### PR TITLE
Document the current aliasing rules for `Box<T>`.

### PR DESCRIPTION
**Pull request by Noratrieb**
Currently, `Box<T>` gets `noalias`, meaning it has the same rules as `&mut T`. This is sparsely documented, even though it can have quite a big impact on unsafe code using box. Therefore, these rules are documented here, with a big warning that they are not normative and subject to change, since we have not yet committed to an aliasing model and the state of `Box<T>` is especially uncertain.

If you have any suggestions and improvements, make sure to leave them here. This is mostly intended to inform people about what is currently going on (to prevent misunderstandings such as [Jon Gjengset's Box aliasing](https://www.youtube.com/watch?v=EY7Wi9fV5bk)). 

This is supposed to _only document current UB_ and not add any new guarantees or rules.

### CODE REVIEW

**Comment by Nilstrieb** [comment]
r? rust-lang/libs-api @rustbot label +T-libs-api -T-libs

**Review by dtolnay — CHANGES_REQUESTED** [review]
I think it needs to be more clear that nothing in this alarming-looking section is relevant to a causal reader who isn't writing their own unsafe code (which is 99%+ of people learning about `Box` from its rustdoc).

Perhaps a section like `# Considerations for unsafe code` would be more appropriate, saying "raw pointer" instead of "pointer", and being explicit like "For more guidance on working with Box from unsafe code, see ucg#326".

File: library/alloc/src/boxed.rs [hunk 1]
```diff
@@ -122,7 +122,21 @@
...
 //! described in [rust-lang/unsafe-code-guidelines#198][ucg#198].
 //!
+//! # Considerations for unsafe code
+//!
+//! **Warning: This section is not normative and is subject to change, possibly
+//! being relaxed in the future! It is a simplified summary of the rules
+//! currently implemented in the compiler.**
+//!
+//! The aliasing rules for `Box<T>` are the same as for `&mut T`. `Box<T>`
+//! asserts uniqeness over its content. Using raw pointers derived from a box
```
Comment (dtolnay) [inline library/alloc/src/boxed.rs:132]:
```suggestion
//! asserts uniqueness over its content. Using raw pointers derived from a box
```

**Review by dtolnay — CHANGES_REQUESTED** [review]

**Review by dtolnay — APPROVED** [review]
Looks good :) thanks!

**Pull request merged**